kairos-contracts/demo-contract/contract/src/main.rs

```diff
@@ -0,0 +1,119 @@
+    RuntimeArgs, URef, U512,
+};
+mod constants;
+use constants::{
+    KAIROS_DEPOSIT_CONTRACT_NAME, KAIROS_DEPOSIT_CONTRACT_PACKAGE, KAIROS_DEPOSIT_CONTRACT_UREF,
+    KAIROS_DEPOSIT_PURSE, KAIROS_LAST_PROCESSED_DEPOSIT_COUNTER, RUNTIME_ARG_AMOUNT,
+    RUNTIME_ARG_TEMP_PURSE,
+};
+mod entry_points;
+mod utils;
+use utils::errors::DepositError;
+use utils::events::Deposit;
+use utils::get_immediate_caller;
+
+// This entry point is called once when the contract is installed
+// and sets up the security badges with the installer as an admin or the
+// optional list of admins.
+// The optional list of admins is passed to the installation session as a runtime argument.
+// The contract purse will be created in contract context so that it is "owned" by the contract
+// rather than the installing account.
+#[no_mangle]
+pub extern "C" fn init() {
+    if runtime::get_key(KAIROS_DEPOSIT_PURSE).is_some() {
+        runtime::revert(DepositError::AlreadyInitialized);
+    }
+
+    // initialize event schema
+    let schemas = Schemas::new().with::<Deposit>();
+    casper_event_standard::init(schemas);
+
+    let new_deposit_purse: URef = system::create_purse();
+    runtime::put_key(KAIROS_DEPOSIT_PURSE, new_deposit_purse.into());
+}
+
+#[no_mangle]
+pub extern "C" fn get_purse() {
+    let deposit_purse: URef = runtime::get_key(KAIROS_DEPOSIT_PURSE)
+        .unwrap_or_revert_with(DepositError::MissingKeyDepositPurse)
+        .into_uref()
+        .unwrap_or_revert();
+    let reference_to_deposit_purse_with_restricted_access =
+        deposit_purse.with_access_rights(AccessRights::ADD);
+    runtime::ret(
+        CLValue::from_t(reference_to_deposit_purse_with_restricted_access)
+            .unwrap_or_revert_with(DepositError::FailedToReturnContractPurseAsReference),
+    );
+}
+
+// Entry point called by a user through session code to deposit funds.
+// Due to Casper < 2.0 purse management and access control, it is necessary that
+// a temporary purse is funded and passed to the deposit contract, since this is
+// the only secure method of making a payment to a contract purse.
+#[no_mangle]
+pub extern "C" fn deposit() {
+    let temp_purse: URef = runtime::get_named_arg(RUNTIME_ARG_TEMP_PURSE);
+    let amount: U512 = runtime::get_named_arg(RUNTIME_ARG_AMOUNT);
+    let deposit_purse_uref: URef = runtime::get_key(KAIROS_DEPOSIT_PURSE)
+        .unwrap_or_revert_with(DepositError::MissingKeyDepositPurse)
+        .into_uref()
+        .unwrap_or_revert_with(ApiError::UnexpectedKeyVariant);
+    system::transfer_from_purse_to_purse(temp_purse, deposit_purse_uref, amount, None)
+        .unwrap_or_revert();
+
+    // kairos utilizes u64 so only amounts that can be converted are accepted.
+    let amount =
+        u64::try_from(amount).unwrap_or_else(|_| runtime::revert(ApiError::InvalidArgument));
+
+    let new_deposit_record: Deposit = Deposit {
+        account: get_immediate_caller().unwrap_or_revert(),
```

Comment:
We only use get_immediate_caller once.
Can we inline it and make it more robust?

Reply:
It is calling `call_stack_element_to_key` which is a function that is quite deeper. I would say it is fine to keep it factored out.